contrib/juliac/abi_export.jl

```diff
@@ -0,0 +1,219 @@
+        indented_println("{")
+        indented_println("  \"id\": ", type_id, ",")
+        indented_println("  \"kind\": \"struct\",")
+        indented_println("  \"name\": ", type_name_json(dt), ",")
+        indented_println("  \"size\": ", Core.sizeof(dt), ",")
+        indented_println("  \"alignment\": ", Base.datatype_alignment(dt), ",")
+        indented_println("  \"fields\": [")
+        for i = 1:Base.datatype_nfields(dt)
+            emit_field_info!(ctx, dt, i; indent = indent + 4)
+            println(ctx.io, i == Base.datatype_nfields(dt) ? "" : ",")
+        end
+        indented_println("  ]")
+        print(ctx.io, " " ^ indent, "}")
+    end
+end
+
+function emit_primitive_type!(ctx::TypeEmitter, @nospecialize(dt::DataType); indent::Int = 0)
+    type_id = ctx.type_ids[dt]
+    let indented_println(args...) = println(ctx.io, " " ^ indent, args...)
+        indented_println("{")
+        indented_println("  \"id\": ", type_id, ",")
+        indented_println("  \"kind\": \"primitive\",")
+        indented_println("  \"name\": ", type_name_json(dt), ",")
+        indented_println("  \"signed\": ", (dt <: Signed), ",")
+        indented_println("  \"bits\": ", 8 * Base.packedsize(dt), ",") # size for reinterpret / in-register
+        indented_println("  \"size\": ", Base.aligned_sizeof(dt), ",") # size with padding / in-memory
+        indented_println("  \"alignment\": ", Base.datatype_alignment(dt))
+        print(ctx.io, " " ^ indent, "}")
+    end
+end
+
+function emit_type_info!(ctx::TypeEmitter, @nospecialize(dt::DataType); indent::Int = 0)
+    if dt.name === Ptr.body.name
+        emit_pointer_info!(ctx, dt; indent)
+    elseif Base.isprimitivetype(dt)
+        emit_primitive_type!(ctx, dt; indent)
+    else
+        emit_struct_info!(ctx, dt; indent)
+    end
+end
+
+function emit_method_info!(ctx::TypeEmitter, method::Core.Method; indent::Int = 0)
+    (rt, sig) = method.ccallable
+    (name, symbol) = let
+        symbol = length(method.ccallable) > 2 ? Symbol(method.ccallable[3]) : method.name
+        iob = IOBuffer()
+        print(IOContext(iob, :print_method_signature_only => true), method)
+        str = String(take!(iob))
+        if symbol !== method.name && startswith(str, String(method.name))
+            # Make a best-effort attempt to use the exported name
+            #
+            # Note: the `startswith` check is to make sure we support 'functor's in arg0,
+            # which Base.@ccallable supports as long as they are singletons.
+            str = replace(str, String(method.name) => String(symbol); count = 1)
+        end
+        (str, String(symbol))
+    end
+
+    argnames = String.(Base.method_argnames(method))
+    let indented_println(args...) = println(ctx.io, " " ^ indent, args...)
+        indented_println("{")
+        indented_println("  \"symbol\": ", escape_string_json(symbol), ",")
+        indented_println("  \"name\": ", escape_string_json(name), ",")
```

Comment:
Maybe call this `julia_signature` instead of `name`? The latter seems too generic.

Reply:
I intentionally kept this a bit generic to (mildly) discourage parsing of the signature

We can allow that of course, but it means we have to maintain an external signature parser and reckon with some outstanding issues, like that `isbits` objects can appear in types and have no canonical printing (https://github.com/JuliaLang/julia/issues/58756)